pennylane/measurements/classical_shadow.py

```diff
@@ -442,6 +442,131 @@ def process_state_with_shots(
+            ]
+        )
+
+        # the diagonalizing matrices corresponding to the Pauli observables above
+        diag_list = np.stack(
+            [
+                qml.Hadamard.compute_matrix(),
+                qml.Hadamard.compute_matrix() @ qml.RZ.compute_matrix(-np.pi / 2),
+                qml.Identity.compute_matrix(),
+            ]
+        )
+        obs = obs_list[recipes]
+        diagonalizers = diag_list[recipes]
+
+        # transpose the state so that the measured wires appear first
+        unmeasured_wires = [i for i in range(num_dev_qubits) if i not in mapped_wires]
+        transposed_state = np.transpose(
+            state,
+            axes=mapped_wires
+            + unmeasured_wires
+            + [w + num_dev_qubits for w in mapped_wires]
+            + [w + num_dev_qubits for w in unmeasured_wires],
+        )
+
+        outcomes = np.zeros((shots, n_qubits))
+        stacked_state = np.repeat(transposed_state[np.newaxis, ...], shots, axis=0)
+
+        for active_qubit in range(n_qubits):
+            # stacked_state loses a dimension each loop
+
+            # trace out every qubit except the first
+            num_remaining_qubits = num_dev_qubits - active_qubit
+            conj_state_first_qubit = ascii_letters[num_remaining_qubits]
+            stacked_dim = ascii_letters[num_remaining_qubits + 1]
+            remaining_dim = ascii_letters[:num_remaining_qubits]
+            traced_dim = remaining_dim[1:]
+
+            state_str = f"{stacked_dim}{remaining_dim}"
+            conj_state_str = f"{conj_state_first_qubit}{traced_dim}"
+            target_str = f"{stacked_dim}a{conj_state_first_qubit}"
+
+            first_qubit_state = np.einsum(
+                f"{state_str}{conj_state_str}->{target_str}",
+                stacked_state,
+            )
+
+            # sample the observables on the first qubit
+            probs = (np.einsum("abc,acb->a", first_qubit_state, obs[:, active_qubit]) + 1) / 2
+            samples = bit_rng.random(size=probs.shape) > probs
+            outcomes[:, active_qubit] = samples
+
+            # collapse the state of the remaining qubits; the next qubit in line
+            # becomes the first qubit for the next iteration
+            U = diagonalizers[:, active_qubit]
+            UT = np.stack([qml.math.conjugate(qml.math.transpose(m)) for m in U])
+
+            # index labeling:
+            # (s, vL, a) (s, a, ..., b, ...) (s, b, vR) -> (s, vL, ..., vR, ...)
+            v_dim_left = ascii_letters[num_remaining_qubits + 2]
+            v_dim_right = ascii_letters[num_remaining_qubits + 3]
+            a_dagger_dim = ascii_letters[num_remaining_qubits + 4]
+
+            U_str = f"{stacked_dim}{v_dim_left}a"
+            rho_str = f"{stacked_dim}a{traced_dim}{a_dagger_dim}..."
+            UT_str = f"{stacked_dim}{a_dagger_dim}{v_dim_right}"
+            new_str = f"{stacked_dim}{v_dim_left}{traced_dim}{v_dim_right}..."
+            rotated_state = np.einsum(
```

Comment:
Could it be worth to figure out the optimal contraction order for this one and hardcode it in?

Reply:
That's a good idea! Let me find some time to figure this out

Reply:
Might be fine not doing that here and now :)

Reply:
Yeah actually I just quickly went through `einsum`'s own `greedy` and `optimal`, both are not introducing very obvious improvement. Probably better we backlog this for future investigation.

Reply:
Their `optimal` is even having performance downgrade which I guess is caused by their own theoretical solver is using comparable time as the whole shadow algorithm.

Reply:
[sc-91553]